HTML API: Complete missing tags in the IN BODY insertion mode.

dmsnell opened the pull request:
Trac ticket: Core-61576

## Summary

 - [x] There are some failing tests that need to pass.
 - [x] This needs a _full_ and careful audit.
 - [x] Full docblock/example review (probably nothing more is needed since everything should be internal or already-documented)

## Description

As part of work to add more spec support to the HTML API, this patch adds support for the remaining missing tags in the IN BODY insertion mode. Not all of the added tags are supported, because in some cases they reset the insertion mode and are reprocessed where they will be rejected.

Included in these changes are some shuffling of the tag rules in IN BODY. This is to make it easier to compare the final code against the HTML specification, so that they appear in the same order as they exist within the spec.

## Testing

This change unlocks 54 new test assertions that have previously been skipped.

html5lib tests
```diff
- Tests: 614, Assertions: 191, Skipped: 423.
+ Tests: 610, Assertions: 275, Skipped: 335.
```

dmsnell commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -581,9 +595,12 @@ public function next_token() {
 			: parent::is_tag_closer();
+
+		// A closing BR tag is treated as if it were a opening tag.
+		return $is_closer_token && 'BR' !== $this->get_tag();
```
Does this belong in the Tag Processor? I don't know. It in some ways is more accurate there, but then it's not possible to distinguish an opening from closing BR tag. Probably it should move into the Tag Processor since it's _never_ possible in HTML to have a closing BR tag.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor-state.php:
```diff
@@ -215,6 +227,31 @@ class WP_HTML_Processor_State {
+	 *
+	 * @var string
+	 */
+	const QUIRKS_MODE = 'quirks-mode';
```
This is interesting. There are 3 document modes:

- "no-quirks"
- "quirks"
- "limited-quirks"

We're missing "limited-quirks".

However, the `compatMode` of a document is "BackCompat" if the document is in quirks mode or "CSS1Compat" if the document is in "no-quirks" or "limited-quirks" modes.

Do you know whether we ignore the difference between no- and limited- quirks modes? Is the plan to add the additional mode when and if it becomes necessary?

dmsnell replied:
So far I haven't seen any implications for limited-quirks mode. Most language reads like "if document is in quirks mode".

I'll double-check, but I think limited quirks comes specifically in the fragment case.

dmsnell replied:
after another scan through the spec, it does seem that limited-quirks mode behaves just like no-quirks mode, and the only impact this has is on case-sensitivity in matching IDs and class names.

I don't see a reason to differentiate them, especially since browsers don't. the name `CSS1Compat` makes more sense to me now too, since it's really about how CSS selectors behave

sirreal replied:
I'm satisfied 👌

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -934,29 +945,149 @@ private function step_in_body() {
+					 *
+					 * This parser does not currently support this behavior.
+					 *
+					 * @todo The spec is unclear on whether this should be ignored. Should it?
```
It seems clear that the only thing to do with this is to add the attributes that aren't already present. Is your question more about whether we should pause here or skip it?

dmsnell replied:
yeah it's about pausing or skipping. I think the only acceptable answer is to skip, but I don't want to assume that without questioning it. if we don't skip it, we'll reach a new BODY opening without a closer, I think.

sirreal commented on src/wp-includes/html-api/class-wp-html-open-elements.php:
```diff
@@ -101,6 +101,49 @@ public function set_push_handler( Closure $handler ) {
+		}
+
+		return null;
+	}
```
This is a nice method, I'd like to be able to use it in both directions by providing a negative index to `walk_up` and a positive index to `walk_down`.

The negative index would make this case much easier:

https://github.com/WordPress/wordpress-develop/blob/aa11864a1eb09d6b12e51e585bdc6dd698c182a4/src/wp-includes/html-api/class-wp-html-processor.php#L1618-L1626

```php
if (
  $this->state->stack_of_open_elements->current_node_is( 'OPTION' ) &&
  'OPTGROUP' === $this->state->stack_of_open_elements->at( -2 )
) {
  // pop
}
```

[This is how `Array.prototype.at` works in JS.](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/at#index)

```suggestion
	/**
	 * Returns the name of the node at the nth position on the stack
	 * of open elements, or `null` if no such position exists.
	 *
	 * Note that this uses a 1-based index, which represents the
	 * "nth item" on the stack, counting from the top, where the
	 * top-most element is the 1st, the second is the 2nd, etc...
	 *
	 * A negative index can be used to access the element beginning
	 * at the bottom of the stack, where the bottom-most element is
	 * `at( -1 )`.
	 *
	 * @todo Skip over "marker" entries on the stack, if appropriate.
	 *
	 * @since 6.7.0
	 *
	 * @param int $nth Retrieve the nth item on the stack, with 1 being
	 *                 the top element, 2 being the second, etc...
	 *                 If a negative index is provided, the lookup will
	 *                 start from the bottom of the stack.
	 * @return string|null Name of the node on the stack at the given location,
	 *                     or `null` if the location isn't on the stack.
	 */
	public function at( int $nth ) {
		if ( 0 === $nth ) {
			return null; // _doing_it_wrong here?
		}

		if ( 0 > $nth ) {
			foreach ( $this->walk_up() as $item ) {
				if ( 0 === ++$nth ) {
					return $item->node_name;
				}
			}
			return null;
		}

		foreach ( $this->walk_down() as $item ) {
			if ( 0 === --$nth ) {
				return $item->node_name;
			}
		}

		return null;
	}
```

I also added handling for `$nth === 0` which seems good. There shouldn't be risk of an infinite loop, but there's no reason to do any iteration with `0`.

---

**Aside observation**

I spent a bit of time reviewing some language which sounded off to me. It seems like the HTML standard inverts conventional stack terminology. The stack of open elements is a LIFO stack, but it starts at the top and pushes to and pops from the bottom, so the terminology used in the HTML API aligns with that:

> The [html](https://html.spec.whatwg.org/multipage/semantics.html#the-html-element) node, however it is created, is the topmost node of the stack. It only gets popped off the stack when the parser [finishes](https://html.spec.whatwg.org/multipage/parsing.html#stop-parsing).
>
> The current node is the bottommost node in this [stack of open elements](https://html.spec.whatwg.org/multipage/parsing.html#stack-of-open-elements).

dmsnell replied:
I do want to be careful about conflating the language from other programming languages and their array data structures and the stack in the HTML spec. maybe `at()` is the wrong name. I originally added this before when I needed to implement another behavior.

perhaps we call it `stack_position( $nth )`

I like your idea for negative numbers, though maybe we could make it even more explicit.

```php
'OPTGROUP' === $this->state->stack_of_open_elements->at_position( 'from-bottom', 2 );
```

sirreal replied:
It's true this could be confusing to use a single method and negative or positive indexes to move from either end, it may not be clear which end is which.

Instead of a single method with some better-named flag, let's match the `walk_up` (start at bottom and go towards top) and `walk_down` (start at top and go towards bottom) which already exist and match the specification stack:

```php
'DIV' === $this->state->stack_of_open_elements->nth_from_bottom( 1 ); // analogous to current node on the stack
'HTML' === $this->state->stack_of_open_elements->nth_from_top( 1 ); // always true
```

dmsnell replied:
this is great, but I think it would be best on its own, so maybe we can follow-up with a change that modifies things in a unified commit. with the other major open PRs, perhaps we can address this once we are basically at full support, or at least until the open PRs are closed.

dmsnell replied:
scratch that. I was playing with it and I'm going to incorporate it here. let's see how we go. there's a part of me that doesn't like adding more index-based methods (as I write it out), but given the spec uses this language, that's fine I guess.

dmsnell replied:
It still feels odd, so I'm going to leave this out for now and ask you your thoughts. For most cases of `walk_up()` and `walk_down()` we're referencing another node and not an index, which seems far less likely to be broken by updates to the code. For example, setting `$parent = $->nth_from_bottom(2)` depends on the current node being at position `nth = 1`. With `->walk_up( $node ) as $parent` we don't have that same fragility by exporting the internals of the stack to the HTML Processor.

I'm open to this, but the gains didn't feel as big as I expected when I went through what you proposed and did my own search to find all the cases.

```diff
diff --git a/src/wp-includes/html-api/class-wp-html-open-elements.php b/src/wp-includes/html-api/class-wp-html-open-elements.php
index 0cd223657a..6f76265b09 100644
--- a/src/wp-includes/html-api/class-wp-html-open-elements.php
+++ b/src/wp-includes/html-api/class-wp-html-open-elements.php
@@ -109,8 +109,6 @@ class WP_HTML_Open_Elements {
 	 * "nth item" on the stack, counting from the top, where the
 	 * top-most element is the 1st, the second is the 2nd, etc...
 	 *
-	 * @todo Skip over "marker" entries on the stack, if appropriate.
-	 *
 	 * @since 6.7.0
 	 *
 	 * @param int $nth Retrieve the nth item on the stack, with 1 being
@@ -128,6 +126,53 @@ class WP_HTML_Open_Elements {
 		return null;
 	}
 
+	/**
+	 * Returns the name of the node at the nth position from the bottom on
+	 * the stack of open elements, or `null` if no such position exists.
+	 *
+	 * Note that this uses a 1-based index, which represents the
+	 * "nth item" on the stack, counting from the bottom, where the
+	 * bottom-most (or deepest) element is the 1st, the second is
+	 * the 2nd, etc...
+	 *
+	 * @since 6.7.0
+	 *
+	 * @param int $nth Retrieve the nth item on the stack, with 1 being
+	 *                 the bottom (depeest) element, 2 being the second, etc...
+	 * @return string|null Name of the node on the stack at the given location,
+	 *                     or `null` if the location isn't on the stack.
+	 */
+	public function nth_from_bottom( int $nth ): ?string {
+		if ( $nth <= 0 || $nth > count( $this->stack ) ) {
+			return null;
+		}
+
+		return $this->stack[ count( $this->stack ) - $nth ]->node_name;
+	}
+
+	/**
+	 * Returns the name of the node at the nth position from the top on
+	 * the stack of open elements, or `null` if no such position exists.
+	 *
+	 * Note that this uses a 1-based index, which represents the
+	 * "nth item" on the stack, counting from the top, where the
+	 * top-most element is the 1st, the second is the 2nd, etc...
+	 *
+	 * @since 6.7.0
+	 *
+	 * @param int $nth Retrieve the nth item on the stack, with 1 being
+	 *                 the top element, 2 being the second, etc...
+	 * @return string|null Name of the node on the stack at the given location,
+	 *                     or `null` if the location isn't on the stack.
+	 */
+	public function nth_from_top( int $nth ): ?string {
+		if ( $nth <= 0 || $nth > count( $this->stack ) ) {
+			return null;
+		}
+
+		return $this->stack[ $nth - 1 ]->node_name;
+	}
+
 	/**
 	 * Reports if a node of a given name is in the stack of open elements.
 	 *
diff --git a/src/wp-includes/html-api/class-wp-html-processor.php b/src/wp-includes/html-api/class-wp-html-processor.php
index f66498b19f..0f2135387e 100644
--- a/src/wp-includes/html-api/class-wp-html-processor.php
+++ b/src/wp-includes/html-api/class-wp-html-processor.php
@@ -1211,7 +1211,7 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 			case '+BODY':
 				if (
 					1 === $this->state->stack_of_open_elements->count() ||
-					'BODY' !== $this->state->stack_of_open_elements->at( 2 ) ||
+					'BODY' !== $this->state->stack_of_open_elements->nth_from_top( 2 ) ||
 					$this->state->stack_of_open_elements->contains( 'TEMPLATE' )
 				) {
 					// Ignore the token.
@@ -1237,7 +1237,7 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 			case '+FRAMESET':
 				if (
 					1 === $this->state->stack_of_open_elements->count() ||
-					'BODY' !== $this->state->stack_of_open_elements->at( 2 ) ||
+					'BODY' !== $this->state->stack_of_open_elements->nth_from_top( 2 ) ||
 					false === $this->state->frameset_ok
 				) {
 					// Ignore the token.
@@ -2358,10 +2358,8 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 			case '-OPTGROUP':
 				$current_node = $this->state->stack_of_open_elements->current_node();
 				if ( $current_node && 'OPTION' === $current_node->node_name ) {
-					foreach ( $this->state->stack_of_open_elements->walk_up( $current_node ) as $parent ) {
-						break;
-					}
-					if ( $parent && 'OPTGROUP' === $parent->node_name ) {
+					$parent = $this->state->stack_of_open_elements->nth_from_bottom( 2 );
+					if ( 'OPTGROUP' === $parent ) {
 						$this->state->stack_of_open_elements->pop();
 					}
 				}
@@ -3338,10 +3336,7 @@ class WP_HTML_Processor extends WP_HTML_Tag_Processor {
 	 */
 	public function reset_insertion_mode(): void {
 		// Set the first node.
-		$first_node = null;
-		foreach ( $this->state->stack_of_open_elements->walk_down() as $first_node ) {
-			break;
-		}
+		$first_node = $this->state->stack_of_open_elements->nth_from_top( 1 );
 
 		/*
 		 * > 1. Let _last_ be false.
```

dmsnell replied:
Strangely, with that patch applied I get one more skipped test/one fewer assertions. I'm not sure why, but I'm thinking it must be a logic error, which only adds to my skepticism of replacing semantic relations with array indices.

sirreal replied:
Let's leave it out for now and revisit after this lands, that's fine.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -934,29 +945,149 @@ private function step_in_body() {
+				 *
+				 * This parser does not currently support this behavior.
+				 *
+				 * @todo The spec is unclear on whether this should be ignored. Should it?
```
Whatever we do we should be consistent with how we handle HTML above.

My inclination is that we should _not stop_. It would be good to handle the attributes, but in the interest of providing developers an interface similar to what they'd expect in a DOM tree, it doesn't make sense for these nodes to appear where they're written in HTML because they'll never actually exist there. There's a possible side-effect of them being present, but side-effects are not a reason to stop at `next_tag` or `next_token`.

dmsnell replied:
yeah I think this is the only viable option.

originally some of this was held up by the need to communicate these changes. someone could have been looking for a BODY element with `data-is-dark` on it, and skipped the BODY because _at the time it didn't have that attribute_, but only once parsing the full document do we realize it had it.

I'd still like to handle this by communicating _the retroactive document change_ (a need for adoption and fostering as well), but for now it's probably good enough simply to list it as a limit of this parser.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1165,6 +1254,143 @@ private function step_in_body() {
+				}
+
+				/*
+				 * @todo This may need to be handled in the Tag Processor and turn into
+				 *       a single self-contained tag like TEXTAREA, whose modifiable text
+				 *       is the rest of the input document as plaintext.
```
That seems like a good idea, I was wondering how to handle this. We can figure that out in a follow-up.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1165,6 +1254,143 @@ private function step_in_body() {
+				}
+
+				$this->generate_implied_end_tags();
+				if ( ! $this->state->stack_of_open_elements->current_node_is( $token_name ) ) {
+					// @todo Record parse error: this error doesn't impact parsing.
+				}
```
We could turn this into a comment like we've done elsewhere, there's no reason to check if we're not reporting errors.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1165,6 +1254,143 @@ private function step_in_body() {
+				if ( ! $this->state->stack_of_open_elements->has_element_in_scope( $token_name ) ) {
+					// @todo Report parse error.
+					// Ignore the token.
+					return $this->step();
+				}
+
+				$this->generate_implied_end_tags();
+				if ( ! $this->state->stack_of_open_elements->current_node_is( $token_name ) ) {
+					// @todo Record parse error: this error doesn't impact parsing.
+				}
+				$this->state->stack_of_open_elements->pop_until( $token_name );
+				return true;
+
+			/*
+			 * > An end tag whose tag name is "form"
+			 */
+			case '-FORM':
+				if ( ! $this->state->stack_of_open_elements->contains( 'TEMPLATE' ) ) {
+					$node                      = $this->state->form_element;
+					$this->state->form_element = null;
+
+					/*
+					 * > If node is null or if the stack of open elements does not have node
+					 * > in scope, then this is a parse error; return and ignore the token.
+					 *
+					 * @todo It's necessary to check if the form token itself is in scope, not
+					 *       simply whether any FORM is in scope.
+					 */
+					if (
+						null === $node ||
+						! $this->state->stack_of_open_elements->has_element_in_scope( 'FORM' )
+					) {
+						// Parse error: ignore the token.
+						return $this->step();
+					}
+
+					$this->generate_implied_end_tags();
+					if ( $node !== $this->state->stack_of_open_elements->current_node() ) {
+						// @todo Indicate a parse error once it's possible. This error does not impact the logic here.
+					}
```
This no-op can be a to-do comment.

dmsnell replied:
For now I'd rather follow the existing pattern and avoid commenting out actual code. I Think it will be easier to grasp. We can easily remove them when appropriate.

sirreal replied:
There are other places we just comment the same pattern, we also have several different comment styles for the same thing. we should do a consistency pass but that can happen later.

dmsnell replied:
A consistency pass after merging some of the open and large PRs sounds good.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1165,6 +1254,143 @@ private function step_in_body() {
+			/*
+			 * > An end tag whose tag name is "form"
+			 */
+			case '-FORM':
```
Bookmarking for myself. I've reviewed everything else in this PR and this file up to the `-FORM` handling.

sirreal reviewed (commented):
Leaving notes from a partial review.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1165,6 +1254,143 @@ private function step_in_body() {
+				$this->state->stack_of_open_elements->pop_until( $token_name );
+				return true;
+
+			/*
+			 * > An end tag whose tag name is "form"
+			 */
+			case '-FORM':
```
This form handling isn't working correctly, but I haven't fully understood the expected behavior according to the specification. This needs to be revisited in general.

dmsnell replied:
do you remember what behavior was not working correctly?

sirreal replied:
Yes, although **it is working correctly now**. We discussed via another channel. We settled on bailing in a problematic case because the breadcrumbs are broken and it's unclear at this time how to resolve it:

https://github.com/WordPress/wordpress-develop/blob/d61f297fed884f2126a15fcdd4e0135eca2c5101/src/wp-includes/html-api/class-wp-html-processor.php#L1559-L1562

My confusion was that a mis-nested form end tag does not close out everything to the form, but does just take the form off the stack of open elements. When I commented, there was a breadcrumbs and depth issue (now resolved by bailing).

Here's an example of the confusing situation:

```html
<!-- misnested form closer -->
<form>
  <div>
    BODY > FORM > DIV
    </form>
    BODY > FORM > DIV (still)
  </div>
BODY (FORM was closed from inside the DIV)

<!-- properly nested form closer -->
<form>
  <div>
    BODY > FORM > DIV
  </div>
  BODY > FORM
</form>
BODY
```

dmsnell replied:
Glad that bookmark update solved these. Perhaps the reason I couldn't find a failing test case in 6.6 is simply that we didn't get into the situations where the bug would present itself.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1165,6 +1254,143 @@ private function step_in_body() {
+						// @todo Indicate a parse error once it's possible. This error does not impact the logic here.
+					}
+
+					$this->state->stack_of_open_elements->remove_node( $node );
```
I'm confident that the behavior in this branch is incorrect, but I'm not sure of the problem yet. I don't think we should ever have cases where we remove a node from the middle of the stack like we do here. That would effectively be putting the descendent nodes at two different places in the resulting DOM tree which I don't think can ever happen.

Here's an example of broken behavior with `<form><span>before</form>after`:

![Screenshot 2024-07-05 at 17 33 13](https://github.com/WordPress/wordpress-develop/assets/841763/99cd9bec-8c41-4fb3-881c-78c7cf54063a)

The `</form>` close tag should be completely ignored, but this branch removes the node. Ignore the HTML API tree structure, it's confused because the depth decrements but with a different parent. Look at the breadcrumbs in the tooltip that are `HTML > BODY > SPAN > #text`. The FORM element has been removed but its SPAN child remains.

I'm not sure what's wrong yet. [I went to compare html5ever (servo's tree builder) and the implementation looks very similar.](https://github.com/servo/html5ever/blob/e69b05c849031d6a2837d7d86372ff14b3f4080e/html5ever/src/tree_builder/rules.rs#L457-L489)

dmsnell replied:
> The </form> close tag should be completely ignored

Why do you say this?

> I don't think we should ever have cases where we remove a node from the middle of the stack like we do here.

This happens after generating the implied end tags, thought that doesn't impact the `SPAN`.

From how I read the spec, it very much _should_ remove the node mid-stack _when we have a parse error that there are unclosed elements_.

This happens during the _adoption agency algorithm_ as well.

dmsnell replied:
<img width="593" alt="Screenshot 2024-07-05 at 10 06 39 AM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/79a7537a-fe7b-4489-b0d6-12ee967aae86">

I'm not sure why we get `after` twice, but it's not confusing I think because of the statefulness of this algorithm. the text nodes are still children of the `SPAN`, but by the time they get added, the `FORM` is no longer open, so it doesn't show up in the breadcrumbs.

we'd have to remember the stack of open elements of the parent when adding the child 🤔

dmsnell replied:
I've added an abort condition in [917061b](https://github.com/WordPress/wordpress-develop/pull/6972/commits/917061b9d6369a5711eb16ac2243536bcd112ce0)

the new traversal is this

<img width="607" alt="Screenshot 2024-07-05 at 10 23 00 AM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/fc71df00-dfc0-4e30-b769-6ec94dbaeb78">

we can deal with this particular error condition later

sirreal replied:
You're right, the form close tag is not ignored, but it does take the form off the stack of open elements _out of order_. It's no longer open, but it's children remain open.

This is easier to see in an example you shared elsewhere:

```html
<form>
  <div>In the form</form> still in the form
  </div>
Not in the form
```

HTML API doesn't represent this well now because we treat the stack of open elements as if it were the path up to the document root when constructing breadcrumbs. Often, that works, but not always.

---

I've explored adding Node classes a few times. There are a few situations that might help, and this is one of them. If we worked with Nodes that kept a reference to their parent node, we could construct breadcrumbs from that list. Then when nodes come off the stack of open elements and have no children referencing them, they can be garbage collected.

sirreal replied:
> I've added an abort condition in [917061b](https://github.com/WordPress/wordpress-develop/pull/6972/commits/917061b9d6369a5711eb16ac2243536bcd112ce0)

That's a good approach for now 👍

dmsnell replied:
I'm interested in the forgetful nodes idea. Something good to explore down the line.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1165,6 +1254,143 @@ private function step_in_body() {
+
+					$this->state->stack_of_open_elements->pop_until( 'FORM' );
+					return true;
+				}
```
There are paths above that don't return, we should add a break here. We don't want to fall through. It also prevents `phpcbf` from mangling the next comment and phpcs from complaining.

```suggestion
				}
				break;
```

I've pushed a fix for this.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1165,6 +1254,143 @@ private function step_in_body() {
+
+					$this->generate_implied_end_tags();
+
+					if ( ! $this->state->stack_of_open_elements->current_node_is( 'FORM' ) ) {
+						// @todo Indicate a parse error once it's possible. This error does not impact the logic here.
+					}
```
No-op parse error, this can be a comment.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1211,14 +1437,30 @@ private function step_in_body() {
 
-				$this->close_a_p_element();
+				$this->generate_implied_end_tags();
+
+				if ( ! $this->state->stack_of_open_elements->current_node_is( $token_name ) ) {
+					// @todo Record parse error: this error doesn't impact parsing.
+				}
```
Noop parse error, comment instead.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1282,11 +1540,41 @@ private function step_in_body() {
 			 * >   entry; i.e. act as if this was a "br" start tag token with no attributes, rather
 			 * >   than the end tag token that it actually is.
+			 *
+			 * Note that this is handled elsewhere, so it should not be possible to reach this code.
```
Do we want to update the unsupported message? If it's handled elsewhere and we don't expect to hit this case the message seems to be outdated.

I do still see an unsupported error with `</br>`.

dmsnell replied:
that was an aspirational comment late at night.

as of 9cc67b0 it should only ever report as an opening BR, even in the Tag Processor

sirreal commented on tests/phpunit/tests/html-api/wpHtmlProcessorHtml5lib.php:
```diff
@@ -200,6 +186,13 @@ private static function build_tree_representation( ?string $fragment_context, st
+					// Self-contained tags contain their inner contents as modifiable text.
+					$modifiable_text = $processor->get_modifiable_text();
+					if ( '' !== $modifiable_text ) {
+						$output .= str_repeat( $indent, $indent_level ) . "\"{$processor->get_modifiable_text()}\"\n";
```
Did you intend to re-use `$modifiable_text` here?

```suggestion
						$output .= str_repeat( $indent, $indent_level ) . "\"{$modifiable_text}\"\n";
```

dmsnell replied:
Reused in [a7f884c](https://github.com/WordPress/wordpress-develop/pull/6972/commits/a7f884c48c763d66149ed964c895abb083af127a)

sirreal reviewed (commented):
Partial review, I'll continue.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -1516,11 +1818,80 @@ private function step_in_body() {
+				$this->insert_html_element( $this->state->current_token );
+				return true;
+
+			/*
+			 * > A start tag whose tag name is "noembed"
+			 * > A start tag whose tag name is "noscript", if the scripting flag is enabled
+			 *
+			 * The scripting flag is never enabled in this parser.
+			 */
+			case '+NOEMBED':
```
I'm worried about not handling `+NOSCRIPT` here. If we don't handle it, I think it ends up being handled as "any other start tag" and it does the wrong thing 🙃 

![Screenshot 2024-07-05 at 18 54 42](https://github.com/WordPress/wordpress-develop/assets/841763/0de42c41-af64-4fe6-9cdc-397f12e6bbd2)

dmsnell replied:
I'm glad you asked, but this is correct the way it is. that rule doesn't exist _in this parser_ because this parser does not enable the scripting flag.

the reason the rule is conditional is because _when the scripting flag is enabled_ the NOSCRIPT tag enters RAWTEXT, skipping any nodes that exist inside it. this makes sense, because in a browser where scripting is enabled, you don't want any parsing to occur within that element.

but we don't run JavaScript, and so if we ignore the NOSCRIPT contents then we will prevent people from modifying the fallback content. if you disable JavaScript and re-examine the DOM tree you should find that it matches the HTML Processor's output.

dmsnell replied:
With JavaScript enabled
<img width="257" alt="Screenshot 2024-07-05 at 10 48 31 AM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/4b596b09-efc4-48dd-b0c6-e1f9e7b7a42a">

With JavaScript disabled
<img width="244" alt="Screenshot 2024-07-05 at 10 48 49 AM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/b6bb28e9-9863-461c-ba68-8f60db45385e">

sirreal replied:
You're right, this makes sense as-is 👍

sirreal commented on src/wp-includes/html-api/class-wp-html-processor.php:
```diff
@@ -2481,7 +2869,25 @@ public function class_list() {
+			return $text;
+		}
+
+		if ( '#comment' === $this->get_token_name() ) {
```
Funky comments probably need this handling too?

dmsnell replied:
maybe? I want to audit this because the rules are all different for various contexts and insertion modes, mainly whether null bytes transform into U+FFFD or whether it's stripped away. I'll follow-up with this, but I don't feel like it has to be in this patch

dmsnell replied:
Updated to handle all situations, which I think is right now.

I also moved the logic into the Tag Processor, because it seems like it's capable enough to know whether to skip the newline. Turns out that if it's _anything_ other than a text node immediately following a PRE or LISTING (since TEXTAREA is self-contained, the logic there is clear), then there's no skipping. A simple flag `WP_HTML_Tag_Processor::skip_next_line_feed` suffices.

sirreal reviewed (approved):
I've left a lot of comments, the majority of this is ready to go once you've considered them. I won't be able to review again soon, so I'm leaving a preemptive approval.

dmsnell committed:
HTML API: Complete missing tags in the IN BODY insertion mode.

As part of work to add more spec support to the HTML API, this patch adds
support for the remaining missing tags in the IN BODY insertion mode. Not
all of the added tags are supported, because in some cases they reset the
insertion mode and are reprocessed where they will be rejected.

See Core-61576.

westonruter commented on src/wp-includes/html-api/class-wp-html-open-elements.php:
```diff
@@ -101,6 +101,51 @@ public function set_push_handler( Closure $handler ) {
+	 * @return string|null Name of the node on the stack at the given location,
+	 *                     or `null` if the location isn't on the stack.
+	 */
+	public function at( int $nth ) {
```
One of the methods has a PHP return type hint, so seems like this should be added to the others as well. For example:

```suggestion
	public function at( int $nth ): ?string {
```

westonruter commented on src/wp-includes/html-api/class-wp-html-open-elements.php:
```diff
@@ -101,6 +101,51 @@ public function set_push_handler( Closure $handler ) {
+	 * @param string $node_name Name of node for which to check.
+	 * @return bool Whether a node of the given name is in the stack of open elements.
+	 */
+	public function contains( $node_name ) {
```
```suggestion
	public function contains( $node_name ): bool {
```

westonruter commented on src/wp-includes/html-api/class-wp-html-open-elements.php:
```diff
@@ -101,6 +101,51 @@ public function set_push_handler( Closure $handler ) {
+	 * @param string $node_name Name of node for which to check.
+	 * @return bool Whether a node of the given name is in the stack of open elements.
+	 */
+	public function contains( $node_name ) {
```
Would `is_open` make more sense?

dmsnell replied:
perhaps. the language in the spec tends to be like "if the node is in the stack of open element" so I don't think it prefers one or the other.

`is_open()` makes sense, but when I read `$this->state->stack_of_open_elements->is_open( 'DIV' )` is reads a bit more awkwardly.

no strong thoughts here.

westonruter replied:
Oh, nevermind. I was thinking this was a method on the processor itself.

westonruter commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -2694,7 +2694,8 @@ public function has_self_closing_flag() {
 			self::STATE_MATCHED_TAG === $this->parser_state &&
-			$this->is_closing_tag
+			$this->is_closing_tag &&
+			'BR' !== $this->get_tag()
```
Why this one special case for a void element?

dmsnell replied:
a good call for an explanatory comment. `BR` is special in that a closing tag like `</br>` is always treated like an opening tag, and also as if it has no attributes.

in the Tag Processor we parse attributes but only store them if the tag is an opener. so if we report this here we should _only_ present this as an opening tag and the HTML API won't have to deal with the confusion of treating it as a closer.

at first I had this only in the HTML Processor, since it's a semantic issue and not a lexical one, but then I thought more about it and felt like there's never a reason to expose a BR closer, since they cannot exist.

dmsnell replied:
Noted in 677d03ca4b24efd92cd967838c1b9ffc9cfba634

westonruter commented on src/wp-includes/html-api/class-wp-html-token.php:
```diff
@@ -63,7 +63,7 @@ class WP_HTML_Token {
 	 * Called when token is garbage-collected or otherwise destroyed.
 	 *
-	 * @var callable|null
+	 * @var ?callable
```
Apparently PHPDoc doesn't actually support this nullable syntax, although PHP itself does. 

https://stackoverflow.com/a/50704179/93579

Same goes for below.

sirreal replied:
Thanks for flagging, [[58742]](https://core.trac.wordpress.org/changeset/58742) was landed to fix PHPDoc annotations in many places.

sirreal commented on src/wp-includes/html-api/class-wp-html-active-formatting-elements.php:
```diff
@@ -86,6 +86,22 @@ public function current_node() {
+	 *
+	 * @since 6.7.0
+	 */
+	public function insert_marker() {
```
```suggestion
	public function insert_marker(): void {
```

westonruter replied:
The `void` return type hint was introduced in PHP 7.1, and WP still supports PHP 7.0, so I don't think this can be added.

sirreal replied:
I believe that's changed. [6.6 dropped support for PHP 7.0 and 7.1](https://make.wordpress.org/core/handbook/references/php-compatibility-and-wordpress-versions/):

> In WordPress 6.6:
[Dropped support](https://make.wordpress.org/core/2024/04/08/dropping-support-for-php-7-1/) for PHP 7.0 & 7.1.

westonruter replied:
Ah! Your good 😄

westonruter replied:
The [requirements page](https://wordpress.org/about/requirements/) needs to be updated.

sirreal replied:
Good spot, working on getting it fixed: https://meta.trac.wordpress.org/ticket/7723

sirreal committed:
Add return type annotations to new methods

Co-authored-by: Weston Ruter <westonruter@google.com>

sirreal commented on src/wp-includes/html-api/class-wp-html-open-elements.php:
```diff
@@ -101,6 +101,51 @@ public function set_push_handler( Closure $handler ) {
+	 * @param string $node_name Name of node for which to check.
+	 * @return bool Whether a node of the given name is in the stack of open elements.
+	 */
+	public function contains( $node_name ): bool {
```
The argument here can also be annotated. This should be safe because it's a method on an internal class.

```suggestion
	public function contains( string $node_name ): bool {
```

sirreal committed:
Add argument type annotation to contains method

dmsnell committed:
Note behavioral change in BR tag.

dmsnell committed:
Remove handling for -BR which cannot appear.

dmsnell committed:
Merge branch 'trunk' into html-api/finish-in-body-insertion-mode

dmsnell committed:
Make up for where WPCS is deficient in understanding.

dmsnell committed:
Fix: use updated name for method.

dmsnell committed:
Remove empty file.

dmsnell committed:
Update type annotations which crashed tests.

dmsnell committed:
Fix test issue from merge conflict and un-skip test suites.

dmsnell committed:
Move IN BODY text active format reconstruction after null-only text node check.

dmsnell committed:
Add missing rules in IN BODY

dmsnell committed:
Add clear-active-formatting-elements method from Jon's TABLE PR.

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

dmsnell committed:
Refine get_modifiable_text() to handle null bytes.

dmsnell committed:
Fix modifiable text.

dmsnell committed:
Add new tests for modifiable text transforms.

dmsnell commented:
@westonruter @sirreal I think all of the feedback has been addressed. additionally:
 - `get_modifiable_text()` in the Tag Processor has been updated to properly handle leading newlines and NULL bytes.
 - an additional test asserts this behavior for `get_modifiable_text()`.
 - I've pulled in `clear_up_to_last_marker()` from #6040 (thanks @sirreal).
 - Fixed some mistakes/oversights after additional review.

I might merge this tomorrow, but I'd appreciate any reviews _in post_. This is so exciting: we're there - at the end of IN BODY.

dmsnell committed:
Merge branch 'trunk' into html-api/finish-in-body-insertion-mode

sirreal commented on src/wp-includes/html-api/class-wp-html-open-elements.php:
```diff
@@ -226,7 +266,31 @@ public function has_element_in_specific_scope( string $tag_name, $termination_li
+	 * >   - SVG title
+	 *
 	 * @since 6.4.0
+	 * @since 6.7.0 Full implementation.
```
Foreign content is not covered yet.
```suggestion
	 * @since 6.7.0 Implementation covering all HTML elements.
```

dmsnell replied:
good thought: clarified in 0c9c40f26c8c8749cadf1563cb51357d12852898

sirreal commented on src/wp-includes/html-api/class-wp-html-open-elements.php:
```diff
@@ -237,23 +301,34 @@ public function has_element_in_scope( string $tag_name ): bool {
+	 *
 	 * @since 6.4.0
 	 * @since 6.5.0 Implemented: no longer throws on every invocation.
+	 * @since 6.7.0 Full implementation.
```
Foreign content is not covered yet.
```suggestion
	 * @since 6.7.0 Implementation covering all HTML elements.
```

dmsnell replied:
good thought: clarified in 0c9c40f26c8c8749cadf1563cb51357d12852898

sirreal commented on src/wp-includes/html-api/class-wp-html-open-elements.php:
```diff
@@ -264,43 +339,88 @@ public function has_element_in_list_item_scope( string $tag_name ): bool {
+	 * >   - button in the HTML namespace
+	 *
 	 * @since 6.4.0
+	 * @since 6.7.0 Full implementation.
```
Foreign content is not covered yet.
```suggestion
	 * @since 6.7.0 Implementation covering all HTML elements.
```

dmsnell replied:
good thought: clarified in 0c9c40f26c8c8749cadf1563cb51357d12852898

sirreal commented on src/wp-includes/html-api/class-wp-html-open-elements.php:
```diff
@@ -540,7 +660,16 @@ public function after_element_push( WP_HTML_Token $item ): void {
 		 * cases where the precalculated value needs to change.
 		 */
 		switch ( $item->node_name ) {
```
Let's add a since note to this method.

dmsnell replied:
I'm less convinced on these two methods since they aren't spec algorithms and since nothing noticeable from the outside should have changed. before, the HTML Processor shouldn't have allowed getting into any of these tags, but now it does, and things continue to work as expected.

does that make any sense?

sirreal replied:
You're right, I agree in both places.

sirreal commented on src/wp-includes/html-api/class-wp-html-open-elements.php:
```diff
@@ -573,11 +702,17 @@ public function after_element_pop( WP_HTML_Token $item ): void {
 		 * cases where the precalculated value needs to change.
 		 */
 		switch ( $item->node_name ) {
```
Let's add a since note to this method.

dmsnell replied:
see previous comment: do we have a reason to add a `@since` tag, since nothing about the interface changed?

sirreal replied:
Agreed, not necessary.

sirreal commented on src/wp-includes/html-api/class-wp-html-processor-state.php:
```diff
@@ -368,6 +393,18 @@ class WP_HTML_Processor_State {
+	 *
+	 * @var string
+	 */
+	public $compat_mode = self::NO_QUIRKS_MODE;
```
The mode has potential to be confusing. I suspect the processor will almost always be used in "no-quirks" mode. I do think there's value in making sure these things are clear. I'm a bit concerned we're mixing `mode`/`compatMode` and their respective values.

---

The HTML standard refers to a _mode_ associated with every document:

> Each [document](https://dom.spec.whatwg.org/#concept-document) has an associated… mode ("no-quirks", "quirks", or "limited-quirks")…

Then it follows with this section "for web developers" (which is how you detect the mode, roughly, via DOM APIs):

> document . [compatMode](https://dom.spec.whatwg.org/#dom-document-compatmode)
Returns the string "BackCompat" if document’s [mode](https://dom.spec.whatwg.org/#concept-document-mode) is "quirks"; otherwise "CSS1Compat".

---

The most correct thing may be to rename this property to `mode` and continue using the quirks values because this is from an HTML processing perspective.

dmsnell replied:
my first inclination was to use the term "mode" from the spec, but that term is so generic and conflates with "insertion mode" and I was nervous using it.

this isn't currently exposed to calling code, and the terms come from the DOM spec, as you linked, and not from the HTML spec.

I can change it to `document_mode`, but I'm still nervous that adopting `mode` is going to be confusing, since it plays such a small role overall and stands against `insertion_mode`

sirreal replied:
I have similar concerns about `mode` being a very generic term. I do like `document_mode`, that would be my preference if you agree 👍

sirreal commented on src/wp-includes/html-api/class-wp-html-tag-processor.php:
```diff
@@ -2858,29 +2924,35 @@ public function get_modifiable_text(): string {
-			return substr( $decoded, 1 );
+			$decoded = substr( $decoded, 1 );
 		}
+		$this->skip_next_linefeed = false;
```
This worries me. Is this line necessary or can we just make sure it's reset when advancing to another token?

It's important that this method be idempotent. We call this from the processor and it can't change the state so that when consuming code calls it it's broken. And consuming code should not need to know that it can't call this more than once, this should be fine but wouldn't work correctly:

```php
if ( '' !== $this->get_modifiable_text() ) {
  return "<h1>" . $this->get_modifiable_text() . "</h1>";
}
```

I noticed this both in the tests when debugging and in the html-api-debugger plugin. I always see the incorrect behavior (with this line) but I'm not sure where the method is being called twice in the plugin.

When this line is commented, things seem to work as expected and tests pass.

dmsnell replied:
it's hard to change when advancing to the next token, because that wipes out the flag when moving from LISTING and PRE to `#text`.

I'll explore setting `last_node_name` or something like this. it's a good catch, thank you.

dmsnell replied:
in [1323448](https://github.com/WordPress/wordpress-develop/pull/6972/commits/1323448ce22f6c8811caf7f50858c06df4674309) I've changed this to `$this->skip_newline_at` as a way of declaratively detecting where to ignore the newline. this is more resilient, but I still have concerns about seeking and updating the document.

I'll try to figure out seeking, but as it stands we don't expose a way to change a PRE or LISTING tag into something else, so I will not worry too much about fixing those cases until we allow getting into them.

dmsnell replied:
for now I've simply listed the behavior after seeking as a limitation and warned to seek to the PRE or LISTING, not to the `#text` node. I don't know how to fix this without introducing unseen overhead in the Tag Processor.

do you think this is reasonable enough?

sirreal reviewed (changes requested):
I have some concerns about the `get_modifiable_text` fixes for null characters and omitted newlines.

If we pull that out, I think everything else is ready in this PR and that change can be done in a follow-up.

dmsnell committed:
Clarify @since tag for open elements.

Co-authored-by: Jon Surrell <sirreal@users.noreply.github.com>

dmsnell committed:
Test idempotency of `get_modifiable_text()`.

dmsnell committed:
Use declarative method of tracking ignored newline instead of imperative one.

sirreal reviewed (approved):
Thanks for the work on this.

This is a remaining issue where a newline is removed from a text node that is only a newline, resulting in a text node with no text. The text node should not be visited at all in this case:

```html
<pre>&#x0A;</pre>
```

This PR is big enough and in a good place, I don't mind landing now and handling that known issue in a follow-up.

dmsnell committed:
Note limitation of newline behavior and add test which currently skips.

dmsnell commented:
> Thanks for the work on this.
> 
> This is a remaining issue where a newline is removed from a text node that is only a newline, resulting in a text node with no text. The text node should not be visited at all in this case:
> 
> ```
> <pre>&#x0A;</pre>
> ```
> 
> This PR is big enough and in a good place, I don't mind landing now and handling that known issue in a follow-up.

Thanks @sirreal. I'm puzzled on this one. While there's no child node in the DOM, I find it strange that anyone would consider there to be no text node there. Perhaps I would feel differently about the actual newline byte 0x0A, but intentionally encoding it makes it seem that from a syntactic perspective, someone wanted it to be there.

Let's continue to explore this in the follow-up.

dmsnell committed:
Rename `compat_mode` to `document_mode`

dmsnell committed:
Merge branch 'trunk' into html-api/finish-in-body-insertion-mode

dmsnell committed:
Linter

dmsnell committed:
Update comments

dmsnell committed:
Update comments.

dmsnell committed:
Update comments.

dmsnell commented:
Merged in [[58779]](https://core.trac.wordpress.org/changeset/58779)
https://github.com/wordpress/wordpress-develop/commit/db30ce9c8c1c6ed0c39a3e16b22a82e7a7307c24
